Allow slice indexing of arrays and lists

ConduitDan opened the pull request:
This submission adds slicing as a option for indexing 

for Lists, Arrays and Matrices you can use ranges and lists to specify which elements you want.

Next task it to add ... as an open range.

Are there any other objects that would benefit from slice indexing? sparse matrices? 

This code has been placed in veneer.c near the array handling do  you have better suggestions for its placement?

I also create an arrayerror -> matrixerror function to throw the correct kind of error depending on what the object being indexed is.

softmattertheory commented:
This is *incredibly cool and powerful and a major advancement*! I had loads of fun trying it. Some suggestions: 

1. We're going to want to add additional types to slicing, I think, so I think we need to refactor slightly. Inside getslice and setslicerecursive there are three if/then trees to determine type that will get unmanageable. These respectively: 
- Check bounds (l 488-495)
- Create new object with correct size (l 516-524)
- Copy information (l 540-561) 

I'd recommend refactoring a bit so that that getslice is called with two functions (passed as function pointers): one to create an object (that does bounds checking) and one to copy information. The contents of the if/then trees can be placed into an array, list and matrix versions respectively. This will be faster (the if/then inside a recursive function is expensive) and easier to extend for other collection objects such as Sparse matrices. 

2. Have a small code clean so that there's
- space [_] between if and the brace: if ()_{
- else is on the same line as the brace: } else 
- A comment before getslice to document the protocol for calling it [especially given the above suggestion]. 

3. Add documentation to the docs. I think maybe a new slices topic would be helpful, or you can add a line or two to the array, list and matrix topics.

joshichaitanya3 commented:
Sorry that it took me a while to get this. It looks great! Should be a pretty powerful addition!

I noticed a small thing: While indexing a `Matrix` with only one index, Morpho assumes the second index to be `0` by default. This however doesn't work with the slicing:
```
>var A = Matrix([[0,1,2,3],[4,5,6,7],[8,9,10,11]])   
>print A
[ 0 1 2 3 ]
[ 4 5 6 7 ]
[ 8 9 10 11 ]
>print A[2]
8
>print A[3] // The single index is not treated as an index for the flattened matrix
Error 'MtrxBnds' : Matrix index out of bounds. 
>print A[1..2]
Error 'MtrxBnds' : Matrix index out of bounds.
>print A[1..2,0]
[ 4 ]
[ 8 ]
>var B = Matrix([1,2,3,4,5]) // The single indexing is more intuitive for column matrices
>print B[1]
2
>print B[0..1]
Error 'MtrxBnds' : Matrix index out of bounds.
>print B[0..1,0]
[ 1 ]
[ 2 ]
```                                                             
I don't know what the expected behavior should be here...

Also, re. other objects that could benefit from slice indexing, a naïve thought would be strings?
`var name = "Chaitanya"`
`print name[0] // Expect: C` --> This already works
`print name[0...4] // expect: Chai`

ConduitDan commented:
I have updated the code to work off of a dimension check a construct and a copy fcn provide to the slice function. 
I have done a style pass and added comment preambles to functions
I have added the ability to index a matrix with a single index 
![image](https://user-images.githubusercontent.com/88550879/153657302-77bfd4f0-0c9a-4a94-85e3-7397a5546049.png)
![image](https://user-images.githubusercontent.com/88550879/153657336-92ac0409-38c3-4651-9809-e49f1fa30364.png)


what things besides matrices arrays and list would we like this for? 
Its fairly easy to add slicing functionally to anything now.
